sources/packages/backend/libs/services/src/msfaa-number/msfaa-number-shared.service.ts

```diff
@@ -0,0 +1,269 @@
+        select: {
+          id: true,
+        },
+        where: {
+          disbursementScheduleStatus: DisbursementScheduleStatus.Pending,
+          studentAssessment: {
+            application: { student: { id: studentId } },
+            offering: { offeringIntensity: offeringIntensity },
+          },
+        },
+      });
+      const schedulesIdsToUpdate = schedulesToUpdate.map(
+        (schedule) => schedule.id,
+      );
+      if (schedulesIdsToUpdate.length) {
+        disbursementScheduleRepo.update(schedulesIdsToUpdate, {
+          msfaaNumber: newMSFAANumber,
+          modifier: auditUser,
+          updatedAt: now,
+        });
+      }
+      return newMSFAANumber;
+    });
+  }
+
+  /**
+   * Generates the next number for an MSFAA
+   * considering the offering intensity.
+   * @param offeringIntensity offering intensity.
+   * @returns number to be used for the next MSFAA.
+   */
+  private async consumeNextSequence(
+    offeringIntensity: OfferingIntensity,
+  ): Promise<string> {
+    let nextNumber: number;
+    const sequenceGroupName = `${offeringIntensity}_MSFAA_STUDENT_NUMBER`;
+    await this.sequenceService.consumeNextSequence(
+      sequenceGroupName,
+      async (nextSequenceNumber: number) => {
+        nextNumber = nextSequenceNumber;
+      },
+    );
+    return nextNumber.toString();
+  }
+
+  /**
+   * Get the application information needed to generate a new MSFAA.
+   * @param applicationId reference application for the MSFAA creation.
+   * @returns the application with the data needed for the MSFAA creation.
+   */
+  private async getApplicationForMSFAACreation(
+    applicationId: number,
+  ): Promise<Application> {
+    const application = await this.applicationRepo.findOne({
+      select: {
+        id: true,
+        applicationStatus: true,
+        student: {
+          id: true,
+        },
+        currentAssessment: {
+          id: true,
+          offering: {
+            offeringIntensity: true,
+          },
+          disbursementSchedules: {
+            id: true,
+            disbursementScheduleStatus: true,
+            msfaaNumber: {
+              id: true,
+              cancelledDate: true,
+            },
+          },
+        },
+      },
+      relations: {
+        student: true,
+        currentAssessment: {
+          offering: true,
+          disbursementSchedules: {
+            msfaaNumber: true,
+          },
+        },
+      },
+      where: {
+        id: applicationId,
+      },
+    });
+
+    if (!application) {
+      throw new CustomNamedError(
+        `Application id ${applicationId} was not found.`,
+        APPLICATION_NOT_FOUND,
+      );
+    }
+    const nowAllowedApplicationStatuses = [
+      ApplicationStatus.Draft,
+      ApplicationStatus.Cancelled,
+      ApplicationStatus.Overwritten,
+    ];
+    if (nowAllowedApplicationStatuses.includes(application.applicationStatus)) {
```

Comment:
Why didn't we include this condition in the query?

Also the Query can be based on status [Assessment,Enrolment,Completed] which will pick just the eligible one.

Reply:
Related to "_Why didn't we include this condition in the query_?": the idea was to provide a more a specific error instead of a generic "Application not found" for this situation. I believe that in the past we talked about both approaches being acceptable.

About checking for "_Assessment,Enrolment,Completed_", the MSFAA as per the assessment workflow right now can be generated while the application is "In Progress", for instance (unless we change it). The idea was only to provide a certain level of validation and ensure that the offering would be present.